consensus/core/src/base_committer.rs

```diff
@@ -0,0 +1,371 @@
+    /// The length of a wave (minimum 3)
+    pub wave_length: u32,
+    /// The offset used in the leader-election protocol. This is used by the
+    /// multi-committer to ensure that each [`BaseCommitter`] instance elects
+    /// a different leader.
+    pub leader_offset: u32,
+    /// The offset of the first wave. This is used by the pipelined committer to
+    /// ensure that each[`BaseCommitter`] instances operates on a different
+    /// view of the dag.
+    pub round_offset: u32,
+}
+
+impl Default for BaseCommitterOptions {
+    fn default() -> Self {
+        Self {
+            wave_length: DEFAULT_WAVE_LENGTH,
+            leader_offset: 0,
+            round_offset: 0,
+        }
+    }
+}
+
+/// The [`BaseCommitter`] contains the bare bone commit logic. Once instantiated,
+/// the method `try_direct_decide` and `try_indirect_decide` can be called at any
+/// time and any number of times (it is idempotent) to determine whether a leader
+/// can be committed or skipped.
+#[allow(unused)]
+pub struct BaseCommitter {
+    /// The committee information
+    committee: Arc<Committee>,
+    /// The consensus leader schedule to be used to resolve the leader for a
+    /// given round.
+    leader_schedule: LeaderSchedule,
+    /// Keep all block data
+    block_store: BlockStore,
+    /// The options used by this committer
+    options: BaseCommitterOptions,
+}
+
+impl BaseCommitter {
+    pub fn new(
+        committee: Arc<Committee>,
+        leader_schedule: LeaderSchedule,
+        block_store: BlockStore,
+    ) -> Self {
+        Self {
+            committee,
+            leader_schedule,
+            block_store,
+            options: BaseCommitterOptions::default(),
+        }
+    }
+
+    pub fn with_options(mut self, options: BaseCommitterOptions) -> Self {
+        assert!(options.wave_length >= MINIMUM_WAVE_LENGTH);
+        self.options = options;
+        self
+    }
+
+    /// Return the wave in which the specified round belongs. This takes into
+    /// account the round offset for when pipelining is enabled.
+    fn wave_number(&self, round: Round) -> WaveNumber {
+        round.saturating_sub(self.options.round_offset) / self.options.wave_length
+    }
+
+    /// Return the leader round of the specified wave. The leader round is always
+    /// the first round of the wave. This takes into account round offset for when
+    /// pipelining is enabled.
+    fn leader_round(&self, wave: WaveNumber) -> Round {
+        wave * self.options.wave_length + self.options.round_offset
+    }
+
+    /// Return the decision round of the specified wave. The decision round is
+    /// always the last round of the wave. This takes into account round offset
+    /// for when pipelining is enabled.
+    fn decision_round(&self, wave: WaveNumber) -> Round {
+        let wave_length = self.options.wave_length;
+        wave * wave_length + wave_length - 1 + self.options.round_offset
```

Comment:
It might be just me, but I would prefer to add parentheses around `(wave * wave_length)` to avoid any bugs on a potential refactoring. Although simple I've seen in the past many times ops accidentally get wrong by such changes.

Reply:
Same comment for `leader_round` method